test/quark-core/QuarkWallet.t.sol

```diff
@@ -297,121 +298,152 @@ contract QuarkWalletTest is Test {
-    //     vm.resumeGasMetering();
-    //     vm.expectEmit(true, true, true, true);
-    //     emit ClearNonce(address(aliceWallet), op.nonce);
-    //     aliceWallet.executeQuarkOperation(op, v, r, s);
-    //     assertEq(counter.number(), 3);
-    //     // can replay the same operation...
-    //     aliceWallet.executeQuarkOperation(op, v, r, s);
-    //     assertEq(counter.number(), 6);
-
-    //     // can cancel the replayable nonce...
-    //     vm.pauseGasMetering();
-    //     QuarkWallet.QuarkOperation memory cancelOtherOp = new QuarkOperationHelper().newBasicOpWithCalldata(
-    //         aliceWallet, cancelOtherScript, abi.encodeWithSignature("run(bytes32)", op.nonce), ScriptType.ScriptAddress
-    //     );
-    //     (uint8 cancel_v, bytes32 cancel_r, bytes32 cancel_s) =
-    //         new SignatureHelper().signOp(alicePrivateKey, aliceWallet, cancelOtherOp);
-    //     vm.resumeGasMetering();
-    //     aliceWallet.executeQuarkOperation(cancelOtherOp, cancel_v, cancel_r, cancel_s);
+    function testAllowsForReusedNonceWithChangedScript() public {
+        // gas: disable gas metering except while executing operations
+        vm.pauseGasMetering();
+        bytes memory incrementer = new YulHelper().getCode("Incrementer.sol/Incrementer.json");
 
-    //     // and now you can no longer replay
-    //     vm.expectRevert(abi.encodeWithSelector(QuarkNonceManager.NonceAlreadySet.selector));
-    //     aliceWallet.executeQuarkOperation(op, v, r, s);
-    // }
+        // 1. use nonce to increment a counter
+        (QuarkWallet.QuarkOperation memory op1, bytes32[] memory submissionTokens) = new QuarkOperationHelper()
+            .newReplayableOpWithCalldata(
+            aliceWallet,
+            incrementer,
+            abi.encodeWithSignature("incrementCounter(address)", address(counter)),
+            ScriptType.ScriptAddress,
+            1
+        );
+        (uint8 v1, bytes32 r1, bytes32 s1) = new SignatureHelper().signOp(alicePrivateKey, aliceWallet, op1);
+
+        QuarkWallet.QuarkOperation memory op2 = QuarkWallet.QuarkOperation({
+            nonce: op1.nonce,
+            isReplayable: true,
+            scriptAddress: address(counter),
+            scriptSources: new bytes[](0),
+            scriptCalldata: abi.encodeWithSignature("setNumber(uint256)", 999),
+            expiry: op1.expiry
+        });
+        (uint8 v2, bytes32 r2, bytes32 s2) = new SignatureHelper().signOp(alicePrivateKey, aliceWallet, op2);
+
+        // gas: meter execute
+        vm.resumeGasMetering();
+        aliceWallet.executeQuarkOperation(op1, v1, r1, s1);
+        // incrementer increments the counter thrice
+        assertEq(counter.number(), 3);
+        // when reusing the nonce but changing the script, allow
+        aliceWallet.executeQuarkOperationWithSubmissionToken(op2, submissionTokens[1], v2, r2, s2);
+        // code didn't change counter since it was callcode
+        assertEq(counter.number(), 3);
```

Comment:
Kind of weird way to test this. Ideally, we'd test the side-effect of the second quark operation. If you don't want to wrap around an Ethcall, then consider using the `CounterScript`.

Reply:
What a weird test. I fixed it now in #209 to just call a different script. It was mostly a refactored test that previously had checked we weren't allowed to change script address, but now we can.